text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+pros:
+* Users can continue to use a syntax they are already familiar with
+cons:
+* User provided variables may be evaluated too soon, before buildpack modifications are applied.
+
+## Lifecycle support for `<app>/.exec`
+When we remove support for `<app>/.profile` we could add support for `<app>/.exec` or similar where `<app>/.exec` must implement the `exec.d` interface.
+pros:
+* Users may still dynamically modify the runtime environment without requiring a specific buildpack
+cons:
+* The exec.d interface must be duplicated in both the buildpack and platform API
+* If we modify the interface the same app might behave differently or fail to run on specific platforms depending on which version of the platform API they are using
+* Users must directly implement our less-than-perfectly-intuitive exec.d interface instead of whatever better UX buildpack authors invent.
+
+## Lifecycle support for `<app>.profile`
+We could consider having the lifecycle convert `<app>/.profile` files into exec.d  
+If we do not remove shell logic from the spec, users will continue to find the launcher behavior vexingly complex. Also, there will be no sane path forward for supporting overridable arguments.
+pros:
+* Fewer breaking changes
+* More consistency behavior across platforms/builders
+cons:
+* complexity in the launcher
+* launcher behavior that fails on certain stacks
+* undesirable coupling between the lifecycle and specific shells
+
+## API compat strategy - Explicit Bash Evaluation
+When adding legacy shell processes to `metadata.toml` we could replace the "direct=false" command with the literal command that will be evaluated.
+For example the following script process in `launch.toml`
+```
+[[process]]
+type = "hi"
+command = "echo hello "${WORLD:-world}"
+direct = false
+```
+Could become the following in `metadata.toml`
+```
+[[process]]
+type = "hi"
+command = ["bash", "-c", 'echo hello "${WORLD:-world}"']
+direct = true # we could even potentially remove this from the metadata.toml schema entirely
+```
+
+However, things get very complicated if we explicitly source profiles or shell evaluated args in `command`. The complexity of the resulting commands probably makes this strategy untenable.
+
+# Prior Art
+[prior-art]: #prior-art
+
+The Paketo Java buildpacks have already converted an extensive collection of profile script to exec.d binaries and converted all processes to direct processes in order to support minimal stacks like `io.paketo.stacks.tiny` and to provide users with more intuitive argument handling.
+
+The Procfile interface is supported by Paketo, Heroku, and Google buildpacks, demonstrating that it is possible to have a consistent interface across buildpack ecosystems without building direct support for that interface in the lifecycle. 
```

Comment:
I'm not sure this is winning example. There are 3 different impls of something that should (IMO) be a native feature. The fact that these haven't converged on a single Procfile buildpacks is worrisome and not a pattern I'd like to repeat.

Reply:
I love to see an RFC to provide utility buildpacks as part of the CNB project, that implement extension specs or common conventions.

Reply:
https://github.com/buildpacks/rfcs/pull/175